src/profile-logic/marker-data.js

```diff
@@ -388,23 +388,53 @@ export function correlateIPCMarkers(threads: Thread[]): IPCMarkerCorrelations {
 
   const correlations = new IPCMarkerCorrelations();
   for (const markers of markersByKey.values()) {
-    const sendTid = markers[0] ? markers[0].tid : undefined;
-    const recvTid = markers[4] ? markers[4].tid : undefined;
+    const startEndpointMarker = markers[0];
+    const endEndpointMarker = markers[4];
+    const sendTid = startEndpointMarker ? startEndpointMarker.tid : undefined;
+    const recvTid = endEndpointMarker ? endEndpointMarker.tid : undefined;
     const sharedData: IPCSharedData = {
-      startTime: markers[0] ? markers[0].data.startTime : undefined,
+      startTime: startEndpointMarker
+        ? startEndpointMarker.data.startTime
+        : undefined,
       sendStartTime: markers[1] ? markers[1].data.startTime : undefined,
       sendEndTime: markers[2] ? markers[2].data.startTime : undefined,
       recvEndTime: markers[3] ? markers[3].data.startTime : undefined,
-      endTime: markers[4] ? markers[4].data.startTime : undefined,
+      endTime: endEndpointMarker ? endEndpointMarker.data.startTime : undefined,
       sendTid,
       recvTid,
       sendThreadName: formatThreadName(sendTid),
       recvThreadName: formatThreadName(recvTid),
     };
 
-    for (const m of markers) {
-      if (m !== undefined) {
-        correlations.set(m.tid, m.index, sharedData);
+    const addedThreadIds = new Set();
+    if (startEndpointMarker) {
+      addedThreadIds.add(startEndpointMarker.tid);
+      correlations.set(
+        startEndpointMarker.tid,
+        startEndpointMarker.index,
+        sharedData
+      );
+    }
+    if (endEndpointMarker) {
+      addedThreadIds.add(endEndpointMarker.tid);
+      correlations.set(
+        endEndpointMarker.tid,
+        endEndpointMarker.index,
+        sharedData
+      );
+    }
+
+    if (!startEndpointMarker || !endEndpointMarker) {
+      // If both markers from sender and receiver threads are present, we can
+      // only add the markers of them to the correlations. That way, the middle
+      // markers from IPC I/O threads will not be visible in the main thread and
+      // make the main thread crowded. It's important to add all the threads to
+      // the marker.
+      for (const m of markers.slice(1, 4)) {
+        if (m !== undefined && !addedThreadIds.has(m.tid)) {
```

Comment:
This condition isn't clear to me. Why are we adding these to the correlations only if the tid hasn't been found yet?

Reply:
Yeah, it's not to add the same marker to the correlation multiple times.
For example, these are the 5 marker in a single IPC message (I also added it to the code as a comment):

```
       endpoint   (sender or background thread)
           |      (or main thread in sender process if they are not profiled)
           |
           v
    transferStart (IPC I/O thread in sender process)
           |      (or main thread in sender process if it's not profiled)
           |
           v
     transferEnd  (IPC I/O thread in sender process)
           |      (or main thread in sender process if it's not profiled)
           |
           v
     transferEnd  (IPC I/O thread in receiver process)
           |      (or main thread in receiver process if it's not profiled)
           |
           v
       endpoint   (receiver or background thread)
                  (or main thread in receiver process if they are not profiled)
```

When we have both endpoints, we don't need to add the same markers to the IPC I/O thread (or main if not profiled), we can just add to the sender and receiver. But when one of them missing, It's safe to add it to the IPC I/O (or main) to make sure that they are visible.

But if we add them to the main thread multiple times, we are going to see the same marker multiple times in a thread (this is the bug I'm solving). 
So we are only adding them to a thread if we didn't add them already.